manta-trusted-setup/src/groth16/test/bn254/ppot.rs

```diff
@@ -0,0 +1,1125 @@
+    __: PhantomData<S>,
+}
+
+impl<S, const POWERS: usize> Size for PerpetualPowersOfTauCeremony<S, POWERS> {
+    const G1_POWERS: usize = (Self::G2_POWERS << 1) - 1;
+    const G2_POWERS: usize = POWERS;
+}
+
+impl<S, const POWERS: usize> Pairing for PerpetualPowersOfTauCeremony<S, POWERS> {
+    type Scalar = Fr;
+    type G1 = G1Affine;
+    type G1Prepared = <Bn254 as PairingEngine>::G1Prepared;
+    type G2 = G2Affine;
+    type G2Prepared = <Bn254 as PairingEngine>::G2Prepared;
+    type Pairing = Bn254;
+
+    fn g1_prime_subgroup_generator() -> Self::G1 {
+        G1Affine::prime_subgroup_generator()
+    }
+
+    fn g2_prime_subgroup_generator() -> Self::G2 {
+        G2Affine::prime_subgroup_generator()
+    }
+}
+
+impl<S, const POWERS: usize> KzgConfiguration for PerpetualPowersOfTauCeremony<S, POWERS> {
+    type DomainTag = u8;
+    type Challenge = [u8; 64];
+    type Response = [u8; 64];
+    type HashToGroup = KZGBlakeHasher<Self>;
+
+    const TAU_DOMAIN_TAG: Self::DomainTag = 0;
+    const ALPHA_DOMAIN_TAG: Self::DomainTag = 1;
+    const BETA_DOMAIN_TAG: Self::DomainTag = 2;
+
+    fn hasher(domain_tag: Self::DomainTag) -> Self::HashToGroup {
+        Self::HashToGroup { domain_tag }
+    }
+
+    fn response(
+        state: &Accumulator<Self>,
+        challenge: &Self::Challenge,
+        proof: &KzgProof<Self>,
+    ) -> Self::Response {
+        let mut hasher = BlakeHasher::default();
+        for item in &state.tau_powers_g1 {
+            item.serialize_uncompressed(&mut hasher).unwrap();
+        }
+        for item in &state.tau_powers_g2 {
+            item.serialize_uncompressed(&mut hasher).unwrap();
+        }
+        for item in &state.alpha_tau_powers_g1 {
+            item.serialize_uncompressed(&mut hasher).unwrap();
+        }
+        for item in &state.beta_tau_powers_g1 {
+            item.serialize_uncompressed(&mut hasher).unwrap();
+        }
+        state.beta_g2.serialize_uncompressed(&mut hasher).unwrap();
+        hasher.0.update(&challenge);
+        proof
+            .tau
+            .serialize(&mut hasher)
+            .expect("Consuming ratio proof of tau failed.");
+        proof
+            .alpha
+            .serialize(&mut hasher)
+            .expect("Consuming ratio proof of alpha failed.");
+        proof
+            .beta
+            .serialize(&mut hasher)
+            .expect("Consuming ratio proof of beta failed.");
+        into_array_unchecked(hasher.0.finalize())
+    }
+}
+
+impl<T, M, S, const POWERS: usize> Deserializer<T, M> for PerpetualPowersOfTauCeremony<S, POWERS>
+where
+    S: Deserializer<T, M>,
+{
+    type Error = S::Error;
+
+    fn deserialize_unchecked<R>(reader: &mut R) -> Result<T, Self::Error>
+    where
+        R: Read,
+    {
+        S::deserialize_unchecked(reader)
+    }
+
+    fn deserialize_compressed<R>(reader: &mut R) -> Result<T, Self::Error>
+    where
+        R: Read,
+    {
+        S::deserialize_compressed(reader)
+    }
+}
```

Comment:
For this `trait`, since `Self` is wrapping an implementation, you'll need to forward _all_ the methods not just the required ones to have the correct semantics. So make sure to also implement `check` and `deserialize_uncompressed`. Also, make sure to implement them in the same order as written in the trait and mark all the methods as `#[inline]` (these 100% should be marked).